dali/operators/reader/loader/webdataset_loader.cc

```diff
@@ -0,0 +1,371 @@
+  bitmask was_output_set;
+  was_output_set.resize(ext_.size(), false);
+  for (size_t wds_shard_index = 0; wds_shard_index < index_paths_.size(); wds_shard_index++) {
+    unfiltered_samples.resize(0);
+    unfiltered_components.resize(0);
+    detail::wds::ParseIndexFile(unfiltered_samples, unfiltered_components,
+                                index_paths_[wds_shard_index]);
+
+    for (auto& sample : unfiltered_samples) {
+      detail::wds::SampleDesc new_sample{
+          detail::wds::VectorRange<detail::wds::ComponentDesc>(components_, components_.size()),
+          detail::wds::VectorRange<size_t>(empty_outputs_, empty_outputs_.size()), wds_shard_index,
+          sample.line_number};
+
+      size_t start_outputs_index = output_indicies_.size();
+      output_indicies_.reserve(start_outputs_index + ext_.size());
+
+      for (auto& component : sample.components) {
+        component.outputs =
+            detail::wds::VectorRange<size_t>(output_indicies_, output_indicies_.size());
+        for (auto& output : ext_map[component.ext]) {
+          if (!was_output_set[output]) {
+            DALI_ENFORCE(
+                component.size % dtypes_[output].size() == 0,
+                make_string("Error in index file at ", index_paths_[wds_shard_index], " line ",
+                            sample.line_number, " - component size and dtype incompatible"));
+            output_indicies_.push_back(output);
+            component.outputs.num++;
+            was_output_set[output] = true;
+          }
+        }
+        if (component.outputs.num) {
+          components_.push_back(std::move(component));
+          new_sample.components.num++;
+        }
+      }
+
+      if (new_sample.components.num < ext_.size()) {
+        switch (missing_component_behavior_) {
+          case detail::wds::MissingExtBehavior::Empty:
+            for (size_t output = 0; output < ext_.size(); output++) {
+              if (!was_output_set[output]) {
+                empty_outputs_.push_back(output);
+                new_sample.empty_outputs.num++;
+              }
+            }
+            samples_.push_back(new_sample);
+            break;
+          case detail::wds::MissingExtBehavior::Skip:
+            components_.resize(new_sample.components.start);
+            output_indicies_.resize(start_outputs_index);
+            break;
+          case detail::wds::MissingExtBehavior::Raise:
+            DALI_FAIL(make_string("Underful sample detected at ", index_paths_[wds_shard_index],
+                                  " line ", sample.line_number));
+            break;
+          default:
+            break;
+        }
+      } else {
+        samples_.push_back(new_sample);
+      }
+      was_output_set.fill(false);
+    }
+  }
+
+  DALI_ENFORCE(static_cast<int64_t>(samples_.size()) >= static_cast<int64_t>(num_shards_),
+               "Number of shards bigger than the total number of samples");
```

Comment:
Should we enforce that? I think padding might do the right thing if enabled.

Reply:
...or not. I'm not sure..

Reply:
no it doesn't. not enforcing that just causes an infinite loop in MoveToNextShard. Actually, it's a mistake that we're not enforcing that directly from the Loader class, since it is a bug that potentially affects all readers that use MoveToNextShard (meaning almost all of them)

Reply:
So how about adding a separate patch with that to the loader itself?

Reply:
alright